### PR TITLE
Initial proposition for problems refactor.

### PR DESCRIPTION
**Pull request by mstechly**
This is just an initial proposition of how we could handle various problems.
I moved the logic that @alexjuda created for the graph partition problem to a separate file. I'm not sure whether this is the right way of handling it though, open to suggestions.

Another way of handling it is to put all the problems derived from qiskit in one file, treat it as private and expose a public interface in `__init__.py`?

Possible further improvements:
- Refactor `get_random_maxcut_hamiltonians` to `get_random_hamiltonian_for_problem`
- Refactor `get_solution_cut_size` to `get_solution_cost_for_problem` 
- Refactor `solve_maxcut_by_exhaustive_search` to `solve_problem_by_exhaustive_search`
- Obviously we need tests ;)

Potential issues:
- We might want to have 2 different implementations of the same problem – e.g. form MaxCut we have one from `qiskit` and our own. For those simple problems, it's not that big of a deal, but for problems like TSP or Vehicle Routing it matters a lot.
- It looks like `qiskit.aqua` [is getting depreciated](https://github.com/Qiskit/qiskit-aqua/commit/f798ce2cf03b43ac19da6e7dd52223cf0193013a) so perhaps it's worth figuring out a different way to handle it.

### CODE REVIEW

File: steps/qaoa.py [hunk 1]
```diff
@@ -57,8 +61,17 @@ def get_graph_partition_hamiltonian(graph):
...
     save_list(thetas.tolist(), "thetas.json")
+
+
+def get_problem_hamiltonian_from_graph(graph, problem_type):
+    if problem_type == "partition":
+        get_graph_partition_hamiltonian(graph)
+    elif problem_type == "maxcut":
+        get_maxcut_hamiltonian(graph)
+    else:
+        raise ValueError(f"Invalid problem type: {problem_type}.")
```
Comment (simonwa7) [inline steps/qaoa.py:77]: I think this would be better represented as two functions, with typed input/return values:
1. `get_graph_partition_hamiltonian`
2. `get_maxcut_hamiltonian`

Rationale:
- Blocks of conditional logic is never ideal and usually leads to more maintenance cost
- Isolating the functions creates clear separation and easier to unit test (when the day comes)
- If we ever need to add any args to either function, we don't end up tainting the other
- This removes the `problem_type` argument and the function name becomes the key describing which problem type you have

File: src/python/zquantum/qaoa/problems/_qiskit_problem_helpers.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    stable_set,
+)
+from zquantum.core.openfermion import qiskitpauli_to_qubitop
+
+
+def _adjacency_matrix(graph: nx.Graph) -> np.ndarray:
+    return nx.to_numpy_array(graph)
+
+
+def _change_matrix_to_qiskit_convention(weights_matrix: np.ndarray):
```
Comment (simonwa7) [inline src/python/zquantum/qaoa/problems/_qiskit_problem_helpers.py:17]: Maybe this should just be: `flip_matrix_term_ordering` since you could call it repeatedly no? It would also then make sense to call it to both flip from qiskit -> zquantum and zquantum -> qiskit

File: src/python/zquantum/qaoa/problems/_qiskit_problem_helpers.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        )
+    elif problem_type in ["stable_set"]:
+        qiskit_operator, offset = stable_set.get_operator(
+            weight_matrix_in_qiskit_convention
+        )
+
+    openfermion_operator = qiskitpauli_to_qubitop(qiskit_operator)
+    # openfermion's QubitOperator doesn't store the offset, we also don't have any
+    # other convenient place to keep track of it, so we're adding it as a free term.
+    return openfermion_operator + _identity_operator(offset)
```
Comment (simonwa7) [inline src/python/zquantum/qaoa/problems/_qiskit_problem_helpers.py:68]: Similar to the other comment about `get_problem_hamiltonian_from_graph` in steps, I would be partial to making this several different functions as opposed to one large one for the same reasoning presented there.

Comment (mstechly) [reply src/python/zquantum/qaoa/problems/_qiskit_problem_helpers.py:68]: This case is slightly more tricky, as we have the following code which would be duplicated all over the place.
```
weight_matrix = _adjacency_matrix(graph)
weight_matrix_in_qiskit_convention = _change_matrix_to_qiskit_convention(
        weight_matrix
    )
...
openfermion_operator = qiskitpauli_to_qubitop(qiskit_operator)
openfermion_operator + _identity_operator(offset)
```
Not sure what's a more elegant way of doing that.

File: steps/qaoa.py (hunk 1 above)
Comment (mstechly) [reply steps/qaoa.py:77]: For the "production code in our library", I would agree.
But this is step that would be part of a workflow, so I intentionally want to keep it flexible and make it easier for someone to use in a workflow. Cause with this one can just iterate over problem types and doesn't need to write conditional logic for calling various python functions.

Comment (alexjuda) [reply steps/qaoa.py:77]: Let's assume all these functions have the same interface – If they wouldn't we'd have problem in the dispatch code anyway. An Orquestra user wants to swap different Hamiltonian generators. With @mstechly's approach this requires changing `"partition"` workflow constant to `"maxcut"`. But is it much improvement over @simonwa7's approach where the user would simply change `"get_graph_partition_hamiltonian"` to `"get_maxcut_hamiltonian"` in the step function name section?

**Review by alexjuda — CHANGES_REQUESTED** [review]
tl;dr sometimes explicitness and repetitiveness is okay 🙂

File: steps/qaoa.py (hunk 1 above)
Comment (mstechly) [reply steps/qaoa.py:77]: Not sure if that's a valid argument, but for me iterating over arguments to a step in a workflow just feels better than iterating over the python function called.
Meeeh, I really don't know, I'm fine either way ;)

**Comment by mstechly** [comment]
@alexjuda @simonwa7 
I have addressed some of the issues we talked about.

It seems like `qiskit.aqua` depreciation is not a problem yet – according to [the migration guide](https://github.com/Qiskit/qiskit-aqua#migration-guide), we'll just need to change imports once it's done.

**Review by simonwa7 — APPROVED** [review]

**Review by alexjuda — APPROVED** [review]
